src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -2163,7 +2163,8 @@ private unsafe void DispatchToAllListeners(int eventId, EventWrittenEventArgs ev
             for (EventDispatcher? dispatcher = m_Dispatchers; dispatcher != null; dispatcher = dispatcher.m_Next)
             {
                 Debug.Assert(dispatcher.m_EventEnabled != null);
-                if (eventId == -1 || dispatcher.m_EventEnabled[eventId])
+                if (LocalAppContextSwitches.DisableEventListenerFiltering ||
+                    (dispatcher.m_EventEnabled[eventId] && dispatcher.m_Listener.IsEventEnabled(eventCallbackArgs)))
```

Comment:
Is there any way to restructure WriteToAllListeners / DispatchToAllListeners such that we can delay creating the EventWrittenEventArgs until / if it's actually needed the first time?  As it stands, it seems like we're creating the fairly-expensive per-event object even if it'll never be used.

Reply:
I just took a second look and yes, I think I should be able to by querying whether any of the listeners have the events enabled before calling WriteToAllListeners from WriteImpl<T> - this would mean that in the case it is actually enabled by an instance of listener, it needs incur a little bit of more perf cost (since needs to filter it again at the DispatchToAllListeners) but I think it still makes sense to do what you said for the sake of pay-for-play. I'll give that a try. Thanks Stephen!

Reply:
We already have a check "event[eventId].enabledForAnyListener" that precedes the object construction. https://github.com/dotnet/runtime/blob/c6857a7441322927396db9170c10c690baa1bf70/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L1291

Similar to above, if these filters aren't working it would be good to understand why not and if they should be fixed vs. creating a second filtering implementation.